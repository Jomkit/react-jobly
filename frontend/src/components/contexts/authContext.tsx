import { createContext } from 'react';
import { authInterface } from '../../types';

export const authContext = createContext<authInterface>({} as authInterface);