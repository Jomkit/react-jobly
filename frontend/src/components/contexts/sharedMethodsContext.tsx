import { createContext } from 'react';
import { sharedMethodsInterface } from '../../types';

export const sharedMethodsContext = createContext<sharedMethodsInterface>({} as sharedMethodsInterface);