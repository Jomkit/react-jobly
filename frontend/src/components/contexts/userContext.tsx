import { createContext } from "react";
import { userInterface } from "../../types";

export const userContext = createContext<userInterface | null>(null);