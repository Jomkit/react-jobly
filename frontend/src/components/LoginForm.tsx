import React, { useContext } from 'react';
import { authInterface, userInterface } from '../types';
import { Form, Formik } from 'formik';
import TextInput from './formikComponents/TextInput';
import { authContext } from './contexts/authContext';
import { useNavigate } from 'react-router-dom';

const LoginForm: React.FC<{}> = () => {
    const { login } = useContext<authInterface>(authContext);
    const navigate = useNavigate();
    
    const initialValues: Partial<userInterface> = {
        username: '',
        password: ''
    }
    
  return (
    <>
        <h1 className="mt-4">Login</h1>
        <Formik
            initialValues={initialValues}
            onSubmit={((values, { setSubmitting }) => {
                login(values);
                setSubmitting(false);
                navigate("/");
            })}
        >
            <Form className="d-flex flex-column col col-6 m-auto p-5 pt-3 border border-3 rounded border-secondary">
                <TextInput label="Username" name='username' type="text" />
                <TextInput label="Password" name='password' type="password" />

                <button className="btn btn-primary mt-2" type="submit">Submit</button>
            </Form>
            
        </Formik>
    </>
  )
}

export default LoginForm