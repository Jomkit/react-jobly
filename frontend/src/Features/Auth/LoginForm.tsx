import React, { useContext } from 'react';
import { sharedMethodsInterface, userInterface } from '../../types.ts';
import { Form, Formik } from 'formik';
import TextInput from '../../components/formikComponents/TextInput';
import { sharedMethodsContext } from '../../components/contexts/sharedMethodsContext';
import { useNavigate } from 'react-router-dom';

const LoginForm: React.FC<{}> = () => {
    const { login } = useContext<sharedMethodsInterface>(sharedMethodsContext);
    const navigate = useNavigate();
    
    const initialValues: Partial<userInterface> = {
        username: '',
        password: ''
    }
    
  return (
    <>
        <h1 className="mt-4 text-white">Login</h1>
        <Formik
            initialValues={initialValues}
            onSubmit={((values, { setSubmitting }) => {
                login(values);
                setSubmitting(false);
                navigate("/");
            })}
        >
            <Form className="d-flex flex-column col col-6 m-auto p-5 pt-3 bg-white border border-3 border-black rounded">
                <TextInput label="Username" name='username' id='username' type="text" />
                <TextInput label="Password" name='password' id='password' type="password" />

                <button className="btn btn-primary mt-2" type="submit">Submit</button>
            </Form>
            
        </Formik>
    </>
  )
}

export default LoginForm