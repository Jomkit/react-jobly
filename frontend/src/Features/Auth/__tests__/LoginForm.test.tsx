import { MemoryRouter } from "react-router-dom";
import { sharedMethodsContext } from "../../../components/contexts/sharedMethodsContext";
import LoginForm from "../LoginForm";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { TESTUSER2 } from "../../../__tests__/_testCommon";

function setup(tsx: React.ReactNode) {
    return {
        user: userEvent.setup(),
        ...render(tsx)
    }
}

const signup = vi.fn();
const login = vi.fn();
const logout = vi.fn();

const TestLoginForm = () => {
    return (
        <MemoryRouter initialEntries={['/']}>
            <sharedMethodsContext.Provider value={{signup, login, logout}}>
                <LoginForm />
            </sharedMethodsContext.Provider>
        </MemoryRouter>
    )
}

it('should render without crashing', () => { 
    render(<TestLoginForm />);
})

it('should match snapshot', () => {
    const { asFragment } = render(<TestLoginForm />);

    expect(asFragment()).toMatchSnapshot();
})

it('should update name field on change', async () => { 
    const { getByLabelText, user } = setup(<TestLoginForm />);

    const nameInput = getByLabelText('Username');
    await user.type(nameInput, 'foo');
    expect(nameInput).toHaveValue('foo');
})

it('should redirect to home page on successful submit', async () => {
    const { getByLabelText, getByText, user } = setup(<TestLoginForm />);

    const userNameInput = getByLabelText('Username');
    const passwordInput = getByLabelText('Password');
    const submitBtn = getByText('Submit');

    await user.type(userNameInput, TESTUSER2.username);
    await user.type(passwordInput, TESTUSER2.password);
    
    await user.click(submitBtn);

    expect(login).toHaveBeenCalled();
    expect(login).toHaveBeenCalledWith({username: TESTUSER2.username, password: TESTUSER2.password});
})