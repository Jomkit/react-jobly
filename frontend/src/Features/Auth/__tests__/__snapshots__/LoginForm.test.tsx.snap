// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <h1
    class="mt-4 text-white"
  >
    Login
  </h1>
  <form
    action="#"
    class="d-flex flex-column col col-6 m-auto p-5 pt-3 bg-white border border-3 border-black rounded"
  >
    <label
      for="username"
    >
      Username
    </label>
    <input
      class="text-input"
      id="username"
      name="username"
      type="text"
      value=""
    />
    <label
      for="password"
    >
      Password
    </label>
    <input
      class="text-input"
      id="password"
      name="password"
      type="password"
      value=""
    />
    <button
      class="btn btn-primary mt-2"
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
